compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -672,7 +672,13 @@ where
         let root = ct.root(tcx);
         debug!(?root);
         match root {
-            Node::Leaf(_) => ControlFlow::CONTINUE,
+            Node::Leaf(ct) => {
+                if let Ok(Some(ac)) = AbstractConst::from_const(tcx, ct) {
+                    recurse(tcx, ac, f)
+                } else {
```

Comment:
in this case we don't want to call `f` for the constant. So we have to first check for `Node::Leaf? , then do the `f(ct)` call.

Reply:
What do you mean? Didn't we already explicitly check for whether it was a leaf previously? Or do you mean the nested abstract const?

Reply:
so rn we always call `f(ct)` on the `Node`. But if we have a `Node::Leaf` which is a nested abstract const, the caller should never see that `Node::Leaf`.

so for `N + 1` with `N = M * 2` the closure `f` should only be called with
```rust
Node::Binop(BinOp::Add, _, _) // N + 1
Node::Binop(BinOp::Mul, _, _) // M * 2
Node::Leaf(ty::ConstKind::Param(M)) // M
Node::Leaf(ty::ConstKind::Value(2)) // 2
Node::Leaf(ty::ConstKind::Value(1)) // 1 
```
but your current impl calls the closure with `Node::Leaf(ty::ConstKind::Unevaluated(M * 2))` as well:
```rust
Node::Binop(BinOp::Add, _, _) // N + 1
Node::Leaf(ty::ConstKind::Unevaluated(M * 2)) // N (which then gets replaced with `M * 2`)
Node::Binop(BinOp::Mul, _, _) // M * 2
Node::Leaf(ty::ConstKind::Param(M)) // M
Node::Leaf(ty::ConstKind::Value(2)) // 2
Node::Leaf(ty::ConstKind::Value(1)) // 1 
```

does that make sense to you?

Reply:
that change made things worse :sweat_smile: you now only visit the following nodes:
```rust
Node::Binop(BinOp::Add, _, _) // N + 1
Node::Leaf(ty::ConstKind::Unevaluated(M * 2)) // N (which then gets replaced with `M * 2`)
Node::Binop(BinOp::Mul, _, _) // M * 2
Node::Leaf(ty::ConstKind::Value(1)) // 1 
```

Reply:
ah yes this makes sense, I can change it tmrw, when it is no longer 3 am